# Notebook for Week-3 Assignment for Capstone Project Course

## Data Scraping

In [1]:
#Importing Pandas for data cleaning
import pandas as pd
import numpy as np

In [2]:
#Importing the data from the downloaded csv file
neighdata = pd.read_csv('TorontoNeighData.csv')
neighdata.head()

Postal code           Borough                Neighborhood
0         M1A      Not assigned                         NaN
1         M2A      Not assigned                         NaN
2         M3A        North York                   Parkwoods
3         M4A        North York            Victoria Village
4         M5A  Downtown Toronto  Regent Park / Harbourfront

In [3]:
#Removing rows with Unassined boroughs
ndata = neighdata[neighdata.Borough != 'Not assigned']
ndata.reset_index(drop=True, inplace=True)
ndata.rename(columns={"Postal code":"PostalCode"}, inplace=True)
ndata.head()

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\frame.py:4133: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


PostalCode           Borough                                  Neighborhood
0        M3A        North York                                     Parkwoods
1        M4A        North York                              Victoria Village
2        M5A  Downtown Toronto                    Regent Park / Harbourfront
3        M6A        North York             Lawrence Manor / Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [4]:
#Checking for duplicate Boroughs
duplicateRowsDF = ndata[ndata.duplicated()]
#df = df.groupby('Name').agg({'Sid':'first', 'Use_Case': ', '.join,'Revenue':'first' }).reset_index()
print(duplicateRowsDF)

   PostalCode Borough Neighborhood
90        NaN     NaN          NaN


In [5]:
#Dropping the row with no data
ndata.dropna(inplace=True)
ndata.head()
#ndata.shape

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


PostalCode           Borough                                  Neighborhood
0        M3A        North York                                     Parkwoods
1        M4A        North York                              Victoria Village
2        M5A  Downtown Toronto                    Regent Park / Harbourfront
3        M6A        North York             Lawrence Manor / Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [6]:
#Converting the / to comma for the neighborhoods in the same Postal code
ndata['Neighborhood'] = ndata['Neighborhood'].str.replace('/',',')
ndata.head()

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


PostalCode           Borough                                  Neighborhood
0        M3A        North York                                     Parkwoods
1        M4A        North York                              Victoria Village
2        M5A  Downtown Toronto                    Regent Park , Harbourfront
3        M6A        North York             Lawrence Manor , Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park , Ontario Provincial Government

In [7]:
#Checking rows with no b=neighborhood data
no_neigh = ndata[ndata['Neighborhood'].isnull()]
#null_data = df[df.isnull().any(axis=1)]
no_neigh.head()

Empty DataFrame
Columns: [PostalCode, Borough, Neighborhood]
Index: []

In [8]:
#Since there are no rows with empty neighborhood, the data cleaning is complete
ndata.shape

(103, 3)

### *****There are 103 rows in the final cleaned dataframe*****

## Appending the geocode information to the dataframe

In [9]:
#Importing the geocodes from csv file
geocodes = pd.read_csv('Geospatial_Coordinates.csv')
geocodes.rename(columns={'Postal Code':'PostalCode'}, inplace=True)
geocodes.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

In [10]:
#Merging the dataframes as a left join
ndata_geo = pd.merge(left=ndata, right=geocodes, on = 'PostalCode')
print(ndata_geo.head())
print(ndata_geo.shape)

  PostalCode           Borough                                  Neighborhood  \
0        M3A        North York                                     Parkwoods   
1        M4A        North York                              Victoria Village   
2        M5A  Downtown Toronto                    Regent Park , Harbourfront   
3        M6A        North York             Lawrence Manor , Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park , Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494  
(103, 5)


## K-means clustering of neighborhoods  within Toronto

In [17]:
#Ccreating dataframe with latitude and longitude information to be used for clustering
xdata = ndata_geo[['Borough', 'Neighborhood', 'Latitude', 'Longitude']]
xdata = xdata[xdata['Borough'].str.contains('Toronto')]
#Setting Borough as the index
#xdata.set_index('Borough', inplace = True)
xdata.reset_index(drop = True,inplace=True)
xdata.shape

(39, 4)

In [18]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(xdata['Borough'].unique()),
        xdata.shape[0]
    )
)

The dataframe has 4 boroughs and 39 neighborhoods.


In [21]:
from geopy.geocoders import Nominatim
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [23]:
import folium
# create map of New York using latitude and longitude values
map_totonto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(xdata['Latitude'], xdata['Longitude'], xdata['Borough'], xdata['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_totonto)  
    
map_totonto

### Obtaining the venues data for Foursquare 

In [24]:
CLIENT_ID = 'UF34A12MOICJUZKMRFTTEBYTYBEU0I1QX0IAVKP2NSD5LKC3' # your Foursquare ID
CLIENT_SECRET = 'DQLGEK1UA5CDKWKUX2ZH2X5A234V0HUP4OY1VVPV2OGCXD5G' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: UF34A12MOICJUZKMRFTTEBYTYBEU0I1QX0IAVKP2NSD5LKC3
CLIENT_SECRET:DQLGEK1UA5CDKWKUX2ZH2X5A234V0HUP4OY1VVPV2OGCXD5G


In [25]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [30]:
import json
import requests
LIMIT=100
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [31]:
toronto_venues = getNearbyVenues(names=xdata['Neighborhood'],
                                   latitudes=xdata['Latitude'],
                                   longitudes=xdata['Longitude']
                                  )

Regent Park , Harbourfront
Queen's Park , Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond , Adelaide , King
Dufferin , Dovercourt Village
Harbourfront East , Union Station , Toronto Islands
Little Portugal , Trinity
The Danforth West , Riverdale
Toronto Dominion Centre , Design Exchange
Brockton , Parkdale Village , Exhibition Place
India Bazaar , The Beaches West
Commerce Court , Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park , The Junction South
North Toronto West
The Annex , North Midtown , Yorkville
Parkdale , Roncesvalles
Davisville
University of Toronto , Harbord
Runnymede , Swansea
Moore Park , Summerhill East
Kensington Market , Chinatown , Grange Park
Summerhill West , Rathnelly , South Hill , Forest Hill SE , Deer Park
CN Tower , King and Spadina , Railway Lands , Harbourfront West , Bathurst
Rosedale
Stn A PO Boxes
St. James Town , Cab

In [32]:
print(toronto_venues.shape)
toronto_venues.head()

(1683, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park , Harbourfront               43.65426              -79.360636   
1  Regent Park , Harbourfront               43.65426              -79.360636   
2  Regent Park , Harbourfront               43.65426              -79.360636   
3  Regent Park , Harbourfront               43.65426              -79.360636   
4  Regent Park , Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4      Morning Glory Cafe       43.653947       -79.361149   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4       Breakfast Spot

In [33]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            57   
Brockton , Parkdale Village , Exhibition Place                         23   
Business reply mail Processing CentrE                                  18   
CN Tower , King and Spadina , Railway Lands , H...                     16   
Central Bay Street                                                     76   
Christie                                                               17   
Church and Wellesley                                                   79   
Commerce Court , Victoria Hotel                                       100   
Davisville                                                             38   
Davisville North                                                        6   
Dufferin , Dovercourt Village                                          16   
First Canadian Place , Underground city                               100   
Forest Hill North & West                                                4   
Garden District, Ryerson                                              100   
Harbourfront East , Union Station , Toronto Isl...                    100   
High Park , The Junction South                                         24   
India Bazaar , The Beaches West                                        19   
Kensington Market , Chinatown , Grange Park                            75   
Lawrence Park                                                           3   
Little Portugal , Trinity                                              50   
Moore Park , Summerhill East                                            1   
North Toronto West                                                     19   
Parkdale , Roncesvalles                                                13   
Queen's Park , Ontario Provincial Government                           34   
Regent Park , Harbourfront                                             46   
Richmond , Adelaide , King                                            100   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede , Swansea                                                    39   
St. James Town                                                        100   
St. James Town , Cabbagetown                                           48   
Stn A PO Boxes                                                         97   
Studio District                                                        41   
Summerhill West , Rathnelly , South Hill , Fore...                     15   
The Annex , North Midtown , Yorkville                                  22   
The Beaches                                                             4   
The Danforth West , Riverdale                                          42   
Toronto Dominion Centre , Design Exchange                             100   
University of Toronto , Harbord                                        35   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             57   
Brockton , Parkdale Village , Exhibition Place                          23   
Business reply mail Processing CentrE                                   18   
CN Tower , King and Spadina , Railway Lands , H...                      16   
Central Bay Street                                                      76   
Christie                                                                17   
Church and Wellesley                                                    79   
Commerce Court , Victoria Hotel                                        100   
Davisville                                                              38   
Davisville North                    

In [35]:
print('There are {} unique venues in Toronto.'.format(len(toronto_venues['Venue Category'].unique())))

There are 233 unique venues in Toronto.


### Processing the venues information for the neighborhoods

In [37]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]
print(toronto_onehot.shape)
toronto_onehot.head()

(1683, 233)


Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  \
0            0                  0        0                   0   
1            0                  0        0                   0   
2            0                  0        0                   0   
3            0                  0        0                   0   
4            0                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Toy / Game Store  Trail  Train Station  \
0             0         0  ...                 0      0              0   
1             0         0  ...                 0      0              0   
2             0         0  ...                 0      0              0   
3             0         0  ...                 0      0              0   
4             0         0  ...                 0      0              0   

   Transportation Service  Vegetarian / Vegan Restaurant  Video Game Store  \
0                       0                              0                 0   
1                       0                              0                 0   
2                       0                              0                 0   
3                       0                              0                 0   
4                       0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Wings Joint  Women's Store  
0                      0         0            0              0  
1                      0         0            0              0  
2                      0         0            0              0  
3                      0         0            0              0  
4                      0         0            0              0  

[5 rows x 233 columns]

In [39]:
# Grouping the data a=as per the venues in each neighborhood 
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
print(toronto_grouped.shape)
toronto_grouped

(39, 233)


Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1      Brockton , Parkdale Village , Exhibition Place     0.043478   
2               Business reply mail Processing CentrE     0.055556   
3   CN Tower , King and Spadina , Railway Lands , ...     0.000000   
4                                  Central Bay Street     0.013158   
5                                            Christie     0.000000   
6                                Church and Wellesley     0.025316   
7                     Commerce Court , Victoria Hotel     0.000000   
8                                          Davisville     0.000000   
9                                    Davisville North     0.000000   
10                      Dufferin , Dovercourt Village     0.000000   
11            First Canadian Place , Underground city     0.000000   
12                           Forest Hill North & West     0.000000   
13                           Garden District, Ryerson     0.000000   
14  Harbourfront East , Union Station , Toronto Is...     0.000000   
15                     High Park , The Junction South     0.000000   
16                    India Bazaar , The Beaches West     0.000000   
17        Kensington Market , Chinatown , Grange Park     0.000000   
18                                      Lawrence Park     0.000000   
19                          Little Portugal , Trinity     0.020000   
20                       Moore Park , Summerhill East     0.000000   
21                                 North Toronto West     0.052632   
22                            Parkdale , Roncesvalles     0.000000   
23       Queen's Park , Ontario Provincial Government     0.029412   
24                         Regent Park , Harbourfront     0.021739   
25                         Richmond , Adelaide , King     0.000000   
26                                           Rosedale     0.000000   
27                                           Roselawn     0.000000   
28                                Runnymede , Swansea     0.025641   
29                                     St. James Town     0.000000   
30                       St. James Town , Cabbagetown     0.000000   
31                                     Stn A PO Boxes     0.000000   
32                                    Studio District     0.024390   
33  Summerhill West , Rathnelly , South Hill , For...     0.000000   
34              The Annex , North Midtown , Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                      The Danforth West , Riverdale     0.023810   
37          Toronto Dominion Centre , Design Exchange     0.000000   
38                    University of Toronto , Harbord     0.028571   

    Afghan Restaurant  Airport  Airport Food Court  Airport Lounge  \
0            0.000000   0.0000              0.0000           0.000   
1            0.000000   0.0000              0.0000           0.000   
2            0.000000   0.0000              0.0000           0.000   
3            0.000000   0.0625              0.0625           0.125   
4            0.000000   0.0000              0.0000           0.000   
5            0.000000   0.0000              0.0000           0.000   
6            0.012658   0.0000              0.0000           0.000   
7            0.000000   0.0000              0.0000           0.000   
8            0.000000   0.0000              0.0000           0.000   
9            0.000000   0.0000              0.0000           0.000   
10           0.000000   0.0000              0.0000           0.000   
11           0.000000   0.0000              0.0000           0.000   
12           0.000000   0.0000              0.0000           0.000   
13           0.000000   0.0000              0.0000           0.000   
14           0.000000   0.0000              0.0000           0.000   
15           0.000000   0.0000              0.0000           0.000   
16           0.000000   0.0000              0.0000

In [44]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [63]:
# Finding the 10 most common venues
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1     Brockton , Parkdale Village , Exhibition Place                  Café   
2              Business reply mail Processing CentrE    Light Rail Station   
3  CN Tower , King and Spadina , Railway Lands , ...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar                Bakery           Cheese Shop   
1        Breakfast Spot           Coffee Shop         Grocery Store   
2           Yoga Studio      Recording Studio            Restaurant   
3        Airport Lounge      Airport Terminal         Boat or Ferry   
4                  Café    Italian Restaurant   Japanese Restaurant   

  5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
0                  Café            Restaurant     Seafood Restaurant   
1                   Gym             Pet Store  Performing Arts Venue   
2         Burrito Place  Fast Food Restaurant          Auto Workshop   
3       Harbor / Marina      Sculpture Garden    Rental Car Location   
4        Sandwich Place           Salad Place       Department Store   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0              Beer Bar        Farmers Market       Greek Restaurant  
1             Nightclub    Italian Restaurant           Intersection  
2        Farmers Market                   Spa             Smoke Shop  
3                 Plane                   Bar               Boutique  
4       Bubble Tea Shop       Thai Restaurant           Dessert Shop

### Applying K-means clustering to the venues data for Toronto neighborhoods using 4 clusters

In [64]:
from sklearn.cluster import KMeans
# set number of clusters
kclusters = 4

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 3])

In [61]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = xdata

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Borough                                  Neighborhood   Latitude  \
0  Downtown Toronto                    Regent Park , Harbourfront  43.654260   
1  Downtown Toronto  Queen's Park , Ontario Provincial Government  43.662301   
2  Downtown Toronto                      Garden District, Ryerson  43.657162   
3  Downtown Toronto                                St. James Town  43.651494   
4      East Toronto                                   The Beaches  43.676357   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.360636               0           Coffee Shop                  Park   
1 -79.389494               0           Coffee Shop                 Diner   
2 -79.378937               0        Clothing Store           Coffee Shop   
3 -79.375418               0           Coffee Shop                  Café   
4 -79.293031               2     Health Food Store                 Trail   

  3rd Most Common Venue      4th Most Common Venue 5th Most Common Venue  \
0                Bakery                        Pub    Mexican Restaurant   
1                  Bank                   Beer Bar              Boutique   
2                  Café  Middle Eastern Restaurant   Japanese Restaurant   
3          Cocktail Bar                 Restaurant    Italian Restaurant   
4                   Pub              Women's Store      Department Store   

  6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
0        Breakfast Spot            Restaurant                         Café   
1        Sandwich Place          Burger Joint                Burrito Place   
2       Bubble Tea Shop        Cosmetics Shop                     Tea Room   
3        Clothing Store              Beer Bar                        Diner   
4  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0               Theater      Electronics Store  
1                  Café                   Park  
2     Electronics Store              Bookstore  
3   American Restaurant                 Bakery  
4   Dumpling Restaurant             Donut Shop

In [65]:
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters